ddl/ddl_api.go

```diff
@@ -1462,93 +1494,152 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 			return infoschema.ErrTableNotExists.GenWithStackByArgs(referIdent.Schema, referIdent.Name)
 		}
 	}
-	if is.TableExists(ident.Schema, ident.Name) {
-		err = infoschema.ErrTableExists.GenWithStackByArgs(ident)
-		if s.IfNotExists {
-			ctx.GetSessionVars().StmtCtx.AppendNote(err)
-			return nil
-		}
-		return err
-	}
 
 	// build tableInfo
 	var tbInfo *model.TableInfo
 	if s.ReferTable != nil {
 		tbInfo = buildTableInfoWithLike(ident, referTbl.Meta())
 	} else {
-		tbInfo, err = buildTableInfoWithCheck(ctx, s, schema.Charset, schema.Collate)
+		tbInfo, err = buildTableInfoWithStmt(ctx, s, schema.Charset, schema.Collate)
 	}
-
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	if err = checkTableInfoValidWithStmt(ctx, tbInfo, s); err != nil {
+		return err
+	}
+
+	onExist := OnExistError
+	if s.IfNotExists {
+		onExist = OnExistIgnore
+	}
+
+	return d.CreateTableWithInfo(ctx, schema.Name, tbInfo, onExist, false /*tryRetainID*/)
+}
+
+func (d *ddl) CreateTableWithInfo(
+	ctx sessionctx.Context,
+	dbName model.CIStr,
+	tbInfo *model.TableInfo,
+	onExist OnExist,
+	tryRetainID bool,
+) (err error) {
+	is := d.GetInfoSchemaWithInterceptor(ctx)
+	schema, ok := is.SchemaByName(dbName)
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(dbName)
+	}
+
+	var oldViewTblID int64
+	if is.TableExists(schema.Name, tbInfo.Name) {
+		err = infoschema.ErrTableExists.GenWithStackByArgs(ast.Ident{Schema: schema.Name, Name: tbInfo.Name})
+		switch onExist {
+		case OnExistIgnore:
+			ctx.GetSessionVars().StmtCtx.AppendNote(err)
+			return nil
+		case OnExistReplace:
+			// only CREATE OR REPLACE VIEW is supported at the moment.
+			if tbInfo.View != nil {
+				oldTable, err := is.TableByName(dbName, tbInfo.Name)
+				if err != nil {
+					return err // logic error, a table exists but can't be found :|
```

Comment:
Under what circumstances does this problem occur?  The implementation of `TableByName` and `TableExists ` is similar, but the return value is different

Reply:
it should never occur, provided that the `is` variable is a snapshot rather than a live view of the info schema. Should it be changed to a panic?

Reply:
The implementation of `TableByName` and `TableExists` is similar, but the return value is different.
Then could we use `TableByName` directly in line1534.

Reply:
PTAL @kennytm